stdlib/public/TensorFlow/Ops.swift

```diff
@@ -1587,3 +1504,205 @@ public extension Tensor {
       size: Tensor<Int32>(upperBounds) - lowerBoundsTensor)
   }
 }
+
+public enum TensorSliceIndex : TensorSliceIndexProtocol {
+  case ellipsis
+  case newAxis
+  case squeezeAxis
+  case index(Int32)
+  case range(Range<Int32>, stride: Int32)
+  case closedRange(ClosedRange<Int32>, stride: Int32)
+  case partialRangeFrom(PartialRangeFrom<Int32>, stride: Int32)
+  case partialRangeUpTo(PartialRangeUpTo<Int32>, stride: Int32)
+  case partialRangeThrough(PartialRangeThrough<Int32>, stride: Int32)
+
+  public var sliceIndex: TensorSliceIndex { return self }
+}
+
+public protocol TensorSliceIndexProtocol {
+  var sliceIndex: TensorSliceIndex { get }
+}
+
+extension Int32 : TensorSliceIndexProtocol {
+  public var sliceIndex: TensorSliceIndex { return .index(self) }
+}
+
+extension Int : TensorSliceIndexProtocol {
+  public var sliceIndex: TensorSliceIndex { return .index(Int32(self)) }
+}
+
+extension Range : TensorSliceIndexProtocol where Bound == Int {
+  public var sliceIndex: TensorSliceIndex {
+    return .range(Int32(self.lowerBound)..<Int32(self.upperBound), stride: 1)
+  }
+}
+
+extension ClosedRange : TensorSliceIndexProtocol where Bound == Int {
+  public var sliceIndex: TensorSliceIndex {
+    return .closedRange(Int32(self.lowerBound)...Int32(self.upperBound), stride: 1)
+  }
+}
+
+extension PartialRangeFrom : TensorSliceIndexProtocol where Bound == Int {
+  public var sliceIndex: TensorSliceIndex {
+    return .partialRangeFrom(Int32(self.lowerBound)..., stride: 1)
+  }
+}
+
+extension PartialRangeUpTo : TensorSliceIndexProtocol where Bound == Int {
+  public var sliceIndex: TensorSliceIndex {
+    return .partialRangeUpTo(..<Int32(self.upperBound), stride: 1)
+  }
+}
+
+extension PartialRangeThrough : TensorSliceIndexProtocol where Bound == Int {
+  public var sliceIndex: TensorSliceIndex {
+    return .partialRangeThrough(...Int32(self.upperBound), stride: 1)
+  }
+}
+
+public extension Tensor {
+  struct IndexPath {
+    @usableFromInline
+    let begin, end, strides: Tensor<Int32>
+
+    @usableFromInline
+    let beginMask, endMask, ellipsisMask, newAxisMask, squeezeAxisMask: Int64
+
+    @usableFromInline
+    init(_ indices: [TensorSliceIndex]) {
```

Comment:
To reduce nesting, I'd suggest defining only the stored properties in this nested struct declaration, and everything else in a `public extension Tensor.IndexPath`.